lib/translator/report_luns.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "report_luns.h"
+
+#include "absl/base/casts.h"
+
+namespace translator {
+
+namespace {
+
+// Return the number of valid namespace IDs in a namespace ID list.
+uint32_t GetNsListLength(const nvme_defs::IdentifyNamespaceList& ns_list) {
+  uint32_t size = 0;
+  for (uint32_t i = 0; i < nvme_defs::kIdentifyNsListMaxLength; ++i) {
+    if (ns_list.ids[i] == 0) break;
+    ++size;
+  }
+  return size;
+}
+
+}  // namespace
+
+// Section 4.5
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToNvme(absl::Span<const uint8_t> scsi_cmd,
+                            nvme_defs::GenericQueueEntryCmd& identify_cmd,
+                            uint32_t& alloc_len) {
+  if (scsi_cmd.size() != sizeof(scsi_defs::ReportLunsCommand)) {
```

Comment:
In other PRs we've been delegating the size check to checking the return value of ReadValue() but keeping the same DebugLof("Malformed ReportLuns command")